pkg/controllers/node/ipam.go

```diff
@@ -55,30 +53,45 @@ const (
 	batchUpdateSize = 1000
 )
 
-func registerPrometheusMetrics() {
-	if !metricsRegistered {
-		// Total IP allocations.
-		ipsGauge = prometheus.NewGaugeVec(prometheus.GaugeOpts{
-			Name: "ipam_allocations_per_node",
-			Help: "Number of IPs allocated",
-		}, []string{"node"})
-		prometheus.MustRegister(ipsGauge)
-
-		// Borrowed IPs.
-		borrowedGauge = prometheus.NewGaugeVec(prometheus.GaugeOpts{
-			Name: "ipam_allocations_borrowed_per_node",
-			Help: "Number of allocated IPs that are from non-affine blocks.",
-		}, []string{"node"})
-		prometheus.MustRegister(borrowedGauge)
-
-		// Blocks per-node.
-		blocksGauge = prometheus.NewGaugeVec(prometheus.GaugeOpts{
-			Name: "ipam_blocks_per_node",
-			Help: "Number of blocks in IPAM",
-		}, []string{"node"})
-		prometheus.MustRegister(blocksGauge)
-	}
-	metricsRegistered = true
+func init() {
+	// Total IP allocations.
+	ipsGauge = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "ipam_allocations_per_node",
+		Help: "Number of IPs allocated",
+	}, []string{"node"})
+	prometheus.MustRegister(ipsGauge)
+
+	// Borrowed IPs.
+	borrowedGauge = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "ipam_allocations_borrowed_per_node",
+		Help: "Number of allocated IPs that are from non-affine blocks.",
+	}, []string{"node"})
+	prometheus.MustRegister(borrowedGauge)
+
+	// Blocks per-node.
+	blocksGauge = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "ipam_blocks_per_node",
+		Help: "Number of blocks in IPAM",
+	}, []string{"node"})
+	prometheus.MustRegister(blocksGauge)
+}
+
+type syncerUpdateType string
+
+const (
+	updateTypeNode   syncerUpdateType = "NodeUpdate"
+	updateTypeBlock  syncerUpdateType = "BlockUpdate"
+	updateTypeStatus syncerUpdateType = "StatusUpdate"
+)
+
+// syncerUpdate is a messanger struct used to send updates received from the syncer
+// to the main worker goroutine. We combine all updates from the syncer into a single
+// channel in order to preserve message order.
+type syncerUpdate struct {
+	updateType syncerUpdateType
```

Comment:
Think I'd just make the channel a `chan interface{}` and use a type switch.   If we're going to switch on type, might as well just use the built-in types that Go provides.  (Given this is an internal channel entirely within this object, I don't think there's a risk of garbage getting sent on the channel.)